### ISSUE TITLE
perf: investigate KNN scaling with # of fragments

### ISSUE THREAD

**Issue by wjones127** [lancedb/lance#1353]
For a fixed number of rows, the time to query grows linearly with the number of fragments. It probably should be flatter.

![Screenshot 2023-10-04 at 12 33 48 PM](https://github.com/lancedb/lance/assets/5488879/5bfc839c-9696-4100-b9b6-fd9d7a25bcd1)

Note: these are cold start, so don't include any caching.

**Comment by wjones127** [comment]
## Flamegraph analysis for KNN

I've noticed there's a big chunk of CPU work done in `concat_batches`. It seems like that doesn't properly pre-size `FixedSizeList`, and so we are spending a lot of CPU time doing memcopies to extend the buffer as we build it. This does not seem to be a factor when we only have a single fragment. 

<details><summary>Flamegraph for Flat KNN on 1024 fragments</summary>

![flamegraph_knn_1024_fragments](https://github.com/lancedb/lance/assets/5488879/d605274e-39fe-4be9-8197-2d511339d0dc)
</details> 

<details><summary>Flamegraph for Flat KNN on 1 fragment</summary>

![flamegraph_knn_1_fragment](https://github.com/lancedb/lance/assets/5488879/f899d08a-f663-4105-b311-ab9ad1c2ffe5)
</details>

**Comment by wjones127** [comment]
## Flamegraph analysis for IVF_PQ

There's no obvious blob of CPU work. It does seems like there's a lot of parked Tokio threads, so I think it's fair to say we could achieve better concurrency. I think two candidates are running `take()` on Arrow batches in parallel, and adding more readahead to `TakeExec`.

<details><summary>Flamegraph for Flat KNN on 1024 Fragments</summary>
<p>

![flamegraph_ivf_pq_1024_fragments](https://github.com/lancedb/lance/assets/5488879/9c9698e2-e1fc-4ebc-9cbf-0759ebd2df43)
</p>
</details> 

<details><summary>Flamegraph for Flat KNN on 1 Fragment</summary>
<p>

![flamegraph_ivf_pq_1_fragment](https://github.com/lancedb/lance/assets/5488879/056cd164-99f8-41b5-95ac-1b04ad1d13e4)

</p>
</details>

**Comment by wjones127** [comment]
This turned out to not be scaling with # of fragments, but scaling with # of rows per fragment when it was < 1024. Having < 1024 rows per fragment means IOs are split up into many tiny requests, increasing the IO overhead. Therefore, it's important to compact files so we don't have any with < 1024 rows.
